### ISSUE TITLE
[ARK] Mods install/update

### ISSUE THREAD

**Issue by m0gL1** [GameServerManagers/LinuxGSM#802]
hi everybody,

our community is running an ark survival evolved server using your script on debian and it seems, that an automated mod update extension can be added in an easy way to your existing code, please check:
### COMMAND

_./steamcmd.sh +login [USERNAME] [PASSWORD] +force_install_dir /home/USERNAME/serverfiles/ShooterGame/Content/Mods +workshop_download_item [ARKSERVER-STEAM-ID] [ARK-MOD-ID] validate +quit_
### 2 KNOWN ISSUES
1. _steamcmd.sh_ doesn´t take the forced path into account, it saves the mod at _/home/USERNAME/steamcmd/steamapps/workshop/content/[ARKSERVER-STEAM-ID]/[ARK-MOD-ID]_ instead of forced path _/home/USERNAME/serverfiles/ShooterGame/Content/Mods_
2. can´t find the _[ARK-MOD-ID].mod_ file, but i assume it´s just a simple renaming of _mod.info_, needs to be checked upfront..

looking forward to get your statement, thanks in advance.
kind regards,
m0gL1

**Comment by dgibbs64** [comment]
My knowledge of Ark and how workshop functions is limited. However I welcome pull requests for this

**Comment by DarthDestroyer** [comment]
With 244.3, it should be done! 

See: http://www.ark-survival.net/en/2016/07/08/ark-survival-evolved-patch-244-3-244-4/

It installs and updates the mods they are in game.ini!

I havent tested it!

**Comment by UltimateByte** [comment]
@DarthDestroyer Nice. Will also solve the "update mods functionality".

PS : About start options within the script : Fixed by commits since June, 10th https://github.com/GameServerManagers/LinuxGSM/commits/master/ARKSurvivalEvolved/arkserver

**Comment by Maygus** [comment]
The auto-update in ARK is ok - but not foolproof. Has problems with larger mods (Ark Advance for one).

However can you review these guys: https://github.com/FezVrasta/ark-server-tools and see how they implemented a...
1: server update check
2: mod update check
3: auto-update mechanic to check / announce / update / restart server combining the above features.

Have these functions then fire off a push-bullet notification would be awesome too.

**Comment by Felino41** [comment]
@DarthDestroyer Which Game.ini should I put them in? So far I´ve only put the mods in the GameUserSettings.ini found under /Saved/config/LinuxServer. Now if I open the Game.ini in that folder its just an empty file?

**Comment by DarthDestroyer** [comment]
Windows & Linux Dedicated Servers now support automatic mod download/installation/updating! Just add “-automanagedmods” to your launch commandline and it will do so before the server launches (note for Maps this will not work unless you’re specifying the Workshop ID to load a Mod Map). On Linux Servers, you need to run this script (included with the ARK dedicated server files) once to install SteamCMD: \SteamCMDInstall.sh
You can also specify a manual list of extra Steam Workshop Mods/Maps/TC ID’s to download/install/update in your Game.ini via (with the commandline as normal to actually use them in-game), example:

[ModInstaller]
ModIDS=680481868
ModIDS=682394267
ModIDS=714672786


====
It is from the Link, i posted! 
Like i said, i havent tested it yet... 

But i had your "emoty ini file", too! 
My solution was: chown it to another user and remove write permission to the start / steam user! 
It will work fine and cant change it back to nothing / default... 
I hate Ark for that! It switches faster Servervars then i boxershorts! If it finds a Var that is deprecated or changed, it will reset your ini without a note or backup! 

The permission trick is one way, but ugly to maintain! :-\
But its a workaround! I havent found a Ini Desciption / Editor / Generator with all new changes,  some of them try its best but arent correct everytime... 

Ark is and will be a survival game! Not only for players... :-\
That is a reason, why we switched back to default without mods... :-(

**Comment by Felino41** [comment]
Thank you for your fast and detailed description! Once I get to it I will definetely try everything out although as you said it seems to be a pain to do so... :/
I would love to be able to just play the game on my server without having to worry so much but it seems as of now you can only really do that without mods...

**Comment by DarthDestroyer** [comment]
if the modupdate / install works like posted above, you only have to restart it! But it doesnt check mod updates on the fly! 
For example, you have some mods autoinstalled and it works fine, server restart at 3am (or something)... 
looks good first, but what happens, when the mod gets an update at 4am? The server will update it at the next 3am, okay... But every user will get the new version installed through workshop and isnt able to play on the "old" version... 

like i said, i havent tested it, but that was the pain before! But more work for me! I had to wait for the updates as client, had to start the game once so i got the mods installed / moved, stop the server, upload every file or mod (because it wasnt easy to know which mod got an update) and after that, start the server again... So, you see the old pain... if you have the server to restart only, now. It would help a little bit... But a Restart every hour isnt the best solution! But it seems to be the only way... :-(

**Comment by UltimateByte** [comment]
So in the end, are there any changes we could make to arkserver to make your life easier ?

**Comment by DarthDestroyer** [comment]
@UltimateByte, maybe... Im not deep into the workshop thing, but if it is possible to check for mod updates in periods and restart the server to force update/install of listed mods?

**Comment by Felino41** [comment]
That would indeed be a valuable improvement. In my opinion Mods are a very important part of the game, but alot of the Server Admins are having a hard time manually updating or restarting servers to keep everything up to date so I´m guessing a viable option to automate these kind of things would be very much appreciated!

**Comment by UltimateByte** [comment]
So what changes ? How do you translate that to code ? Or at least what would be the process ? Any idea ?
How would you explain that to somebody like me that doesn't use Arma 3 so that he could maybe add the feature  ?

**Comment by DarthDestroyer** [comment]
OK... Let me try it:

First: Check every x Minutes if there is an Mod Update (listed in the lgsm config?!)
Second: If there is an update => Restart the Server

Yeah, the lgsm config has to write / set the ModList into the ark.ini but this shouldnt be hard...
Or, maybe there is a way to put it in the commandline, too? Instead of writing it into the ini?

But like i said, im not into the workshop thing and if there is a way to check...
I had written a script to download mods... Maybe i find it, Gimme a Sec!

**Comment by UltimateByte** [comment]
Thanks. Yes, it would definitely help to know how the thing works. :)

**Comment by DarthDestroyer** [comment]
Okay... i'm not a scripter at this point! ;)

I had two files, for the mods!

1.
`#!/bin/bash
./steamcmd/steamcmd.sh +runscript /home/steam/ark/mods.txt`

2.
`@ShutdownOnFailedCommand 1
@NoPromptForPassword 1
login anonymous // <username> <password>
// server
force_install_dir /home/steam/ark/serverfiles
// app_update 376030 validate
// workshop
workshop_download_item 346110 609380111 validate
workshop_download_item 346110 543859212 validate
workshop_download_item 346110 564895376 validate
workshop_download_item 346110 531396658 validate
workshop_download_item 346110 554678442 validate
workshop_download_item 346110 510590313 validate
workshop_download_item 346110 544598236 validate
workshop_download_item 346110 520373758 validate
workshop_download_item 346110 531713045 validate
// workshop_download_item 346110 581683725 validate
workshop_download_item 346110 539464369 validate
workshop_download_item 346110 534952636 validate
workshop_download_item 346110 527194172 validate
workshop_download_item 346110 523835795 validate
workshop_download_item 346110 558651608 validate
workshop_download_item 346110 525507438 validate
workshop_download_item 346110 536214294 validate
workshop_download_item 346110 543752418 validate
workshop_download_item 346110 572277026 validate
workshop_download_item 346110 520879363 validate
workshop_download_item 346110 505861851 validate
// workshop_download_item 346110 512261028 validate
workshop_download_item 346110 574069627 validate
workshop_download_item 346110 543828831 validate
quit`

Let me explain my Try! (I have to say Try, because i had a big mistake! The downloaded Mods are "packed" or something else and after this, i couldnt figure out the right way!)

You start the First file and it does the List of Steam Commands, listed in File 2!
I works well and stable... But it doesnt get the "real" files!
If there is a Workshop-Command, to check these Files for an update (there should be a command, the client does the same!) you could check the result and at a difference, you could Restart the Server easily... ( There is no other way to start the mod download and Mod loading :( )

**Comment by UltimateByte** [comment]
Thanks a lot. I'm pretty sure it'll help the one that will dev this. Maybe me. Please note that i'm working on other LGSM stuff currently that needs to be finished first, so this is a low priority for me. As usual, anyone willing to pull-request the code for feature is more than welcome.

**Comment by usud85** [comment]
i have a solution for updates, my server checks automatically for updates .. then downloads everything shuts down ... warns users and startup again

**Comment by zazufira** [comment]
What is your solution? Please

**Comment by usud85** [comment]
i do not know if i can post pieces of 3rd party codes !?

**Comment by DarthDestroyer** [comment]
It depends on the License, but if you got code, it is likely open source... 

Atleast, you could post your source where you found something ;-)

**Comment by usud85** [comment]
i used parts of this tool

https://github.com/FezVrasta/ark-server-tools

it has possibility to update mods and game itself, i warn users with mcrcon

**Comment by zazufira** [comment]
Ok, remains to find the solution ^^,
thanks anyway  :-)

**Comment by UltimateByte** [comment]
Closing, because of https://github.com/GameServerManagers/LinuxGSM/issues/1336
